### PR TITLE
REF/CLN: maybe_downcast_to_dtype

### PR DESCRIPTION
**Pull request by jbrockmendel**
Separate out a numeric-only maybe_downcast_to_dtype_numeric from maybe_downcast_to_dtype.  Avoid a giant try/except by checking the correct things up front.

The non-numeric portion of maybe_downcast_to_dtype casts to datetime64, datetime64tz, or PeriodArray.  Following #27683 the next step in cleaning up internals is going to be getting rid of _try_cast_result, which will involve replacing a usage of maybe_cast_to_dtype with the less-broadly-scoped maybe_cast_to_dtype_numeric.

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -118,83 +120,115 @@ def trans(x):
...
+            result = result.astype(dtype)
+        except Exception:
+            if dtype.tz:
+                # convert to datetime and change timezone
+                from pandas import to_datetime
+
+                result = to_datetime(result).tz_localize("utc")
+                result = result.tz_convert(dtype.tz)
+
+    elif dtype.type == Period:
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: how about isintance instead

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -118,83 +120,115 @@ def trans(x):
...
+    if not isinstance(dtype, np.dtype):
+        # e.g. SparseDtype has no itemsize attr
+        return result
+
+    if isinstance(result, list):
+        # reached via groupoby.agg _ohlc; really this should be handled
+        #  earlier
+        result = np.array(result)
+
+    def trans(x):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: rather than doing this, I would pass in a callable directly

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -118,83 +120,115 @@ def trans(x):
...
+            return PeriodArray(result, freq=dtype.freq)
+        except TypeError:
+            # e.g. TypeError: int() argument must be a string, a
+            #  bytes-like object or a number, not 'Period
+            pass
+
+    return result
+
+
+def maybe_downcast_numeric(result, dtype, do_round: bool = False):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: this is very similar to to_numeric; would plan as a followup to move to_numeric logic here and call this.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: actually i think the `==` should be `is`, will update

File: pandas/core/dtypes/cast.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: good idea

File: pandas/core/dtypes/cast.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: that gives a lot more degrees of freedom to the caller, I'd rather it just be a bool kwarg until/unless we need something more

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks, this is pretty dense code, good to simplify a bit.

**Pull request merged**